src/boosting/sample_strategy.cpp

```diff
@@ -0,0 +1,28 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+
+#include <LightGBM/sample_strategy.h>
+#include "goss.hpp"
+#include "bagging.hpp"
+
+namespace LightGBM {
+
+SampleStrategy* SampleStrategy::CreateSampleStrategy(const Config* config, const Dataset* train_data, const ObjectiveFunction* objective_function, int num_tree_per_iteration) {
+  bool use_goss_as_boosting = config->boosting == std::string("goss");
+  bool use_goss_as_strategy = config->data_sample_strategy == std::string("goss");
+  if (use_goss_as_boosting) {
+    Log::Warning("Setting goss as `boosting_type` is NOT suggested. Please set `data_sample_strategy = goss` in your config file.");
+    if (use_goss_as_strategy) {
+      Log::Warning("Both `boosting_type` and `data_sample_strategy` are set as GOSS. Only one time of sampling will be conducted. Please check and modify your config file.");
+    }
```

Comment:
Could you explain to me why this is problematic? I don't fully understand, but would like to understand a more informative warning message if you could help me understand this a bit better.

I'd like to be able to say something more specific than "please update your config"

Reply:
This warning was added as we worried that if `boosting_type` and `data_sample_strategy` are both set to `goss`

- users might think GOSS would be applied **twice** (sampling `sampled data` again)
- the real `boosting_type` (GBDT) is not indicated explicitly 

 I agree that we should say something more. Do you have any suggestions?

Reply:
Thanks for the explanation!

In that case, I think one warning is sufficient. It seems to me that we just want to say "stop passing `boosting_type="goss"`, that's now a separate option called `sample_strategy`".

What do you think about the following?

> Found boosting_type="goss". For backwards compatibility reasons, LightGBM interprets this as boosting_type="gbdt",data_sample_strategy="goss". To suppress this warning, set data_sample_strategy="goss" instead.

If you agree with me, then please remove the second `if (use_goss_as_strategy) {` warning.

Reply:
I agree this is sufficient to help users figure out our changes! Fixed in [73d7db](https://github.com/GuangdaLiu/LightGBM/commit/73d7db76785a5a4b13bacf4c1026dbeb6ba62b64).